File: typescript/src/client/fingerprint.ts
/**
 * Device fingerprint derivation.
 *
 *   hex(SHA-256( sorted(sources).join('\n') ))
 *
 * Sorting is lexicographic on the raw source strings so that source
 * collection order doesn't change the hash. Sources are newline-joined (no
 * delimiter escaping needed — newlines can't appear in machine-id / MAC /
 * OS-id strings, and we assert it). The application-provided salt is just
 * another source — no special treatment, no concatenation tricks.
 *
 * This module is I/O-free at the core: `fingerprintFromSources` takes an
 * explicit string array and hashes. The `defaultFingerprintSources` helper
 * collects the canonical defaults (OS id, machine-id, MAC/host UUID, salt)
 * and lives separately because it's the I/O layer of the pluggable-sources
 * design.
 */

import { createHash } from 'node:crypto';

/**
 * A fingerprint source contributor. `collect()` returns either a non-empty
 * string that participates in the hash, or `null` if the source is
 * unavailable on this platform (e.g. no machine-id file) — nulls are
 * filtered out before sorting. Throwing is reserved for configuration
 * errors (salt missing, etc.); "couldn't read that file" should be null.
 */
export interface FingerprintSource {
  readonly name: string;
  collect(): Promise<string | null> | string | null;
}

/**
 * Compute a fingerprint from an already-collected list of source strings.
 * Inputs are validated (no empty, no embedded newlines), sorted
 * lexicographically, joined with `\n`, and hashed.
 *
 * This is the pure core. Callers who want to test with synthetic sources
 * or replace the collection layer entirely can bypass {@link collectFingerprint}.
 */
export function fingerprintFromSources(sources: readonly string[]): string {
  if (sources.length === 0) {
    throw new Error('fingerprint requires at least one source string');
  }
  for (const s of sources) {
    if (s.length === 0) {
      throw new Error('fingerprint source strings must be non-empty');
    }
    if (s.includes('\n')) {
      throw new Error('fingerprint source strings must not contain newlines');
    }
  }
  const sorted = [...sources].sort();
  const hash = createHash('sha256');
  hash.update(sorted.join('\n'));
  return hash.digest('hex');
}

/**
 * Collect fingerprint source strings by invoking each {@link FingerprintSource}
 * in parallel. Nulls (source unavailable) are filtered out. If every source
 * returns null, the call throws — a zero-input fingerprint is useless and
 * would trivially collide across machines.
 */
export async function collectFingerprint(sources: readonly FingerprintSource[]): Promise<string> {
  if (sources.length === 0) {
    throw new Error('at least one fingerprint source must be configured');
  }
  const collected = await Promise.all(
    sources.map(async (src) => {
      const raw = await src.collect();
      return raw === null || raw.length === 0 ? null : raw;
    }),
  );
  const strings = collected.filter((v): v is string => v !== null);
  if (strings.length === 0) {
    throw new Error(
      `all configured fingerprint sources returned null — cannot derive a fingerprint`,
    );
  }
  return fingerprintFromSources(strings);
}

// ---------- default sources ----------

/**
 * Build the default source list: OS id, machine-id, primary MAC/host
 * UUID, and the caller-supplied app salt. Each source is a thin wrapper
 * around a platform read; I/O is deferred to `collect()` time so the
 * source list itself is cheap to construct and test.
 *
 * Callers who want a different source mix MUST pass their own array —
 * this helper is the canonical default and is NEVER silently merged
 * with a caller's custom list.
 */
export function defaultFingerprintSources(appSalt: string): readonly FingerprintSource[] {
  if (appSalt.length === 0) {
    throw new Error('appSalt must be a non-empty string — it anchors the fingerprint to your app');
  }
  return [
    { name: 'os.id', collect: () => readOsId() },
    { name: 'machine.id', collect: () => readMachineId() },
    { name: 'net.primaryMac', collect: () => readPrimaryMac() },
    { name: 'app.salt', collect: () => `salt:${appSalt}` },
  ];
}

/** Best-effort OS identifier. Node exposes `process.platform` (e.g. `linux`,
 *  `darwin`, `win32`). Combine with `process.arch` to disambiguate x64 vs
 *  arm64 darwin machines whose IDs would otherwise collide on virtualized
 *  platforms where the MAC is stable but the architecture differs. */
function readOsId(): string {
  return `os:${process.platform}:${process.arch}`;
}

/** Linux: `/etc/machine-id` or `/var/lib/dbus/machine-id`.
 *  macOS: IOPlatformUUID via `ioreg` — requires a shell-out, which we skip
 *    for now; the MAC source is usually sufficient on macOS.
 *  Windows: `MachineGuid` from the registry — likewise skipped for now.
 *
 *  Returns null when unreadable. Spec allows null-degradation as long as
 *  at least one source produces a value. */
async function readMachineId(): Promise<string | null> {
  try {
    const fs = await import('node:fs/promises');
    const candidates = ['/etc/machine-id', '/var/lib/dbus/machine-id'];
    for (const path of candidates) {
      try {
        const raw = await fs.readFile(path, 'utf8');
        const trimmed = raw.trim();
        if (trimmed.length > 0) return `machine:${trimmed}`;
      } catch {
        // Try next candidate.
      }
    }
    return null;
  } catch {
    return null;
  }
}

/** Primary non-internal MAC address from `os.networkInterfaces()`. Picks
 *  the lexicographically-first interface to keep ordering deterministic
 *  across boots (Node's insertion order is not guaranteed stable on all
 *  platforms). Returns null when no external interface has a MAC — e.g.
 *  inside a sandboxed environment. */
async function readPrimaryMac(): Promise<string | null> {
  try {
    const os = await import('node:os');
    const ifaces = os.networkInterfaces();
    const macs: string[] = [];
    for (const [ifaceName, entries] of Object.entries(ifaces)) {
      if (entries === undefined) continue;
      for (const entry of entries) {
        if (entry.internal) continue;
        // Node fills `mac` with '00:00:00:00:00:00' on some virtual ifaces.
        if (entry.mac.length === 0 || entry.mac === '00:00:00:00:00:00') continue;
        macs.push(`${ifaceName}|${entry.mac}`);
      }
    }
    if (macs.length === 0) return null;
    macs.sort();
    return `mac:${macs[0]}`;
  } catch {
    return null;
  }
}

File: licensing/client/fingerprint.go
package client

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"net"
	"os"
	"runtime"
	"sort"
	"strings"
)

// FingerprintSource is a pluggable source of device identity strings.
// Collect returns "" when the source is unavailable on this platform.
type FingerprintSource interface {
	Name() string
	Collect() (string, error)
}

// FingerprintFromSources is a pure function: sort, join with newlines,
// SHA-256, return lowercase hex. Inputs must be non-empty and contain no
// newlines.
func FingerprintFromSources(sources []string) (string, error) {
	if len(sources) == 0 {
		return "", fmt.Errorf("fingerprint requires at least one source")
	}
	for _, s := range sources {
		if s == "" {
			return "", fmt.Errorf("fingerprint source must not be empty")
		}
		if strings.ContainsRune(s, '\n') {
			return "", fmt.Errorf("fingerprint source must not contain newlines")
		}
	}
	sorted := make([]string, len(sources))
	copy(sorted, sources)
	sort.Strings(sorted)
	joined := strings.Join(sorted, "\n")
	h := sha256.Sum256([]byte(joined))
	return hex.EncodeToString(h[:]), nil
}

// CollectFingerprint runs all sources, filters empty results, and hashes.
func CollectFingerprint(sources []FingerprintSource) (string, error) {
	var vals []string
	for _, src := range sources {
		v, err := src.Collect()
		if err != nil {
			return "", fmt.Errorf("fingerprint source %s: %w", src.Name(), err)
		}
		if v != "" {
			vals = append(vals, v)
		}
	}
	if len(vals) == 0 {
		return "", fmt.Errorf("all fingerprint sources returned empty")
	}
	return FingerprintFromSources(vals)
}

// DefaultFingerprintSources returns the standard set of sources matching
// the TS implementation: OS info, machine ID, primary MAC, and app salt.
func DefaultFingerprintSources(appSalt string) []FingerprintSource {
	if appSalt == "" {
		panic("DefaultFingerprintSources: appSalt must not be empty")
	}
	return []FingerprintSource{
		osSource{},
		machineIDSource{},
		primaryMACSource{},
		saltSource{salt: appSalt},
	}
}

// ---------- built-in sources ----------

type osSource struct{}

// Name returns the fingerprint source identifier.
func (osSource) Name() string { return "os.id" }

// Collect returns a `os:<goos>:<goarch>` string derived from runtime build tags.
func (osSource) Collect() (string, error) {
	return fmt.Sprintf("os:%s:%s", runtime.GOOS, runtime.GOARCH), nil
}

type machineIDSource struct{}

// Name returns the fingerprint source identifier.
func (machineIDSource) Name() string { return "machine.id" }

// Collect reads /etc/machine-id (or dbus fallback) and returns a `machine:<id>` string.
// Returns "" when no machine-id file is readable (e.g., non-Linux).
func (machineIDSource) Collect() (string, error) {
	for _, path := range []string{"/etc/machine-id", "/var/lib/dbus/machine-id"} {
		data, err := os.ReadFile(path)
		if err != nil {
			continue
		}
		id := strings.TrimSpace(string(data))
		if id != "" {
			return "machine:" + id, nil
		}
	}
	return "", nil // unavailable on this platform
}

type primaryMACSource struct{}

// Name returns the fingerprint source identifier.
func (primaryMACSource) Name() string { return "net.primaryMac" }

// Collect returns a `mac:<iface>|<hwaddr>` string for the lexicographically
// first non-loopback interface with a non-zero MAC. Returns "" when no usable
// interface is present.
func (primaryMACSource) Collect() (string, error) {
	ifaces, err := net.Interfaces()
	if err != nil {
		return "", nil
	}
	type entry struct {
		name string
		mac  string
	}
	var candidates []entry
	for _, iface := range ifaces {
		if iface.Flags&net.FlagLoopback != 0 {
			continue
		}
		mac := iface.HardwareAddr.String()
		if mac == "" || mac == "00:00:00:00:00:00" {
			continue
		}
		candidates = append(candidates, entry{name: iface.Name, mac: mac})
	}
	if len(candidates) == 0 {
		return "", nil
	}
	// Pick lexicographically first by interface name (matches TS behavior).
	sort.Slice(candidates, func(i, j int) bool {
		return candidates[i].name < candidates[j].name
	})
	c := candidates[0]
	return fmt.Sprintf("mac:%s|%s", c.name, c.mac), nil
}

type saltSource struct {
	salt string
}

// Name returns the fingerprint source identifier.
func (s saltSource) Name() string { return "app.salt" }

// Collect returns the caller-provided app salt prefixed with "salt:" so it
// participates in the fingerprint but stays distinguishable from OS inputs.
func (s saltSource) Collect() (string, error) {
	return "salt:" + s.salt, nil
}
